### PR TITLE
Add a `mhlo.dynamice_slice` converter pass

### PR DESCRIPTION
**Pull request by Orion34-lanbo**
1. add a pass to convert `mhlo.dynamic_slice` with dynamic shape input to `mhlo.real_dynamic_slice`, we have encountered such `mhlo.dynamic_slice` in easy rec's models with sparse input 

3. add a regression test for sparse subgraph

### CODE REVIEW

**Review by qiuxiafei — DISMISSED** [review]
Generally LGTM ~

File: tao_compiler/mlir/disc/transforms/disc_dynamic_slice_converter.cc [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+
+  LogicalResult matchAndRewrite(mhlo::DynamicSliceOp op,
+                                PatternRewriter& rewriter) const override {
+    Location loc = op.getLoc();
+    Value input = op.operand();
+    auto input_type = input.getType().dyn_cast<RankedTensorType>();
+    if (!input_type || input_type.hasStaticShape()) return failure();
+
+    // slice_sizes: i64 attr
+    auto sliceSizes = op.slice_sizes().getValues<int64_t>();
```
[the remark this answers is not in the document]
Comment (Orion34-lanbo) [reply tao_compiler/mlir/disc/transforms/disc_dynamic_slice_converter.cc]: fixed

File: tao_compiler/mlir/disc/transforms/tests/mhlo-dynamic-slice-convert.mlir [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+  // CHECK: %[[V6:.*]] = arith.minsi %[[V5]], %[[V2]] : i64
+  // CHECK: %[[V7:.*]] = arith.addi %[[V6]], %{{.*}} : i64
+  // CHECK: %[[V8:.*]] = tensor.from_elements %[[V6]] : tensor<1xi64>
+  // CHECK: %[[V9:.*]] = tensor.from_elements %[[V7]] : tensor<1xi64>
+  // CHECK: %[[V10:.*]] = mhlo.real_dynamic_slice %[[ARG0:.*]], %[[V8]], %[[V9]], %{{.*}} : (tensor<?xi64>, tensor<1xi64>, tensor<1xi64>, tensor<1xi64>) -> tensor<1xi64>
+  %out = "mhlo.dynamic_slice"(%arg0, %arg1) {
+    slice_sizes = dense<1> : tensor<1xi64>
+  } : (tensor<?xi64>, tensor<i32>) -> tensor<1xi64>
+  return %out : tensor<1xi64>
+}
```
Comment (qiuxiafei) [inline tao_compiler/mlir/disc/transforms/tests/mhlo-dynamic-slice-convert.mlir:20]: need yet another case to test against constant start indeces?

Comment (Orion34-lanbo) [reply tao_compiler/mlir/disc/transforms/tests/mhlo-dynamic-slice-convert.mlir:20]: Added

**Review by qiuxiafei — APPROVED** [review]

**Review by qiuxiafei — APPROVED** [review]

**Pull request merged**